### DISCUSSION TITLE
Add ability for authorize to accept a lambda

### DISCUSSION THREAD

**Discussion by raldred** [palkan/action_policy#198] [Ideas]
Add the ability for authorize to accept a lambda as an authorization target would make it super flexible.
And could be used with custom middleware that uses a thread-isolated attributes singleton like that provided by `ActiveSupport::CurrentAttributes`

Something like...

```
class AuthSingleton < ActiveSupport::CurrentAttributes
   attribute :user
end

class PostPolicy < ActionPolicy::Base
  authorize :auth, -> { AuthSingleton.instance }

  def update?
    auth.user.admin? || (auth.user.id == record.user_id)
  end
end
```

Feels like this could be implemented relatively easily by allowing an additional `proc` option when [initializing authorization_targets](https://github.com/palkan/action_policy/blob/master/lib/action_policy/policy/authorization.rb#L69)

**Comment by palkan** [comment]
That's an interesting idea.

Maybe, we should provide an ability to specify the default for the authorization context:

```rb
authorize :auth, default: proc { AuthSingleton.instance }
# or with a method name
authorize :auth, default: :auth_from_singleton

def auth_from_singleton
  AuthSingleton.instance
end
```

The second option could be useful if you want to override the default handler in a subclass.

WDYT?

**Reply by raldred** [reply]
Either would suffice for our use case.
The later is probably better as you say it allows for subclassing and follows similar patterns you already have with `implicit_authorization_target` and `authorization_strict_namespace`

Maybe instead of `default`, use `from` or `via` for readability

```ruby
authorize :auth, from: :auth_singleton

def auth_singleton
  AuthSingleton.instance
end
```

**Reply by palkan** [reply]
Yeah, I like `via` 👍

**Reply by davidalejandroaguilar** [reply]
I think the `authorize` option you went for is `through`, right? However, we still can't do `authorize :user, through: -> { Current.user }`, unless I'm missing something?

**Reply by raldred** [reply]
No, `through` doesn't provide the functionality required, it wont accept a proc. I suppose my original suggestion could have been to extended `through` to support a proc, rather than introducing something new.

**Reply by davidalejandroaguilar** [reply]
Right. Should we allow `through` to accept a proc then? I can lend a hand if needed.

**Reply by palkan** [reply]
Thanks @davidalejandroaguilar, I'm open for contributions.

The only thing that I'm not sure about is what should this Proc accept? Or what's gonna be an execution context (e.g., `#instance_exec`)?
